### PR TITLE
Add `SparseGPT` pass

### PR DESCRIPTION
**Pull request by jambayk**
## Describe your changes
Add `SparseGPT` pass which performs model pruning proposed in https://arxiv.org/abs/2301.00774. The pass is based on the original implementation at https://github.com/IST-DASLab/sparsegpt. 

Added corresponding example, test and docs. 

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Format your code by running `pre-commit run --all-files`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
Add `SparseGPT` pass for one-shot model pruning on large GPT like models using the algorithm proposed in https://arxiv.org/abs/2301.00774.

### CODE REVIEW

File: olive/passes/pytorch/sparsegpt.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            ),
+            "max_layer": PassConfigParam(
+                type_=int, default_value=None, description="Prune all layers with id < max_layer."
+            ),
+            "layer_name_filter": PassConfigParam(
+                type_=Union[str, List[str]],
+                default_value=None,
+                description="Only prune layers whose name contains the given string(s).",
+            ),
+            "compute_device": PassConfigParam(
```
Comment (guotuofeng) [inline olive/passes/pytorch/sparsegpt.py:65]: could we use the accelerator_spec.device?

Comment (jambayk) [reply olive/passes/pytorch/sparsegpt.py:65]: I consider using but decided against it since this is not the same as `accelerator_spec.device` which is the device we are targeting for inference. 
Here, it is the device we want to use to run the algorithm on. for example, our target device might be cpu but we can still use gpu for this algorithm. It doesn't affect the output model. Just runs faster on gpu if available.

Comment (guotuofeng) [reply olive/passes/pytorch/sparsegpt.py:65]: that make sense. would you please add more comments here?

Comment (jambayk) [reply olive/passes/pytorch/sparsegpt.py:65]: Yes sure! Should I add to the description or just code comment for the dev?

File: docs/source/features/passes/pytorch.md [hunk 2]
```diff
@@ -59,3 +59,24 @@ c. Run QAT training with default training loop.
 
 Check out [this file](https://github.com/microsoft/Olive/blob/main/examples/resnet/user_script.py)
 for an example implementation of `"user_script.py"` and `"create_train_dataloader"`.
+
+## SparseGPT
+`SparseGPT` prunes GPT like models using a pruning method called [SparseGPT](https://arxiv.org/abs/2301.00774). This one-shot pruning method can perform unstructured
```
Comment (guotuofeng) [inline docs/source/features/passes/pytorch.md:64]: are there any accuracy drop we can refer to between the original GPT model and the pruned model? If yes, we can put it here.

File: olive/passes/pytorch/sparsegpt.py (hunk 1 above)
Comment (guotuofeng) [reply olive/passes/pytorch/sparsegpt.py:65]: we can add comments either place and it is up to you. The purpose of the comments is just to describe why we choose different device.

Comment (jambayk) [reply olive/passes/pytorch/sparsegpt.py:65]: Sounds good. Thanks! adding comment above the parameter.

File: docs/source/features/passes/pytorch.md (hunk 2 above)
Comment (jambayk) [reply docs/source/features/passes/pytorch.md:64]: The authors measured perplexity on multiple models, sizes and sparsity so I didn't know what results to report here. 
So I linked the original paper for reference. Do you think that's okay?

Comment (guotuofeng) [reply docs/source/features/passes/pytorch.md:64]: do we have any benchmark data?

Comment (jambayk) [reply docs/source/features/passes/pytorch.md:64]: From Olive? Not yet, I am waiting for access to supported gpus for latency measurements. The only results are those reported in the paper. 
I did get similar perplexity results on a couple of OPT models (2.7B and 4.7B).

File: olive/passes/pytorch/sparsegpt.py (hunk 1 above)
Comment (jambayk) [reply olive/passes/pytorch/sparsegpt.py:68]: Done!

File: docs/source/features/passes/pytorch.md (hunk 2 above)
Comment (jambayk) [reply docs/source/features/passes/pytorch.md:64]: Updated the doc to say "refer to paper for performance results".

Comment (guotuofeng) [reply docs/source/features/passes/pytorch.md:64]: do we have plan to run the benchmark for open llama with this technique?

Comment (jambayk) [reply docs/source/features/passes/pytorch.md:64]: Yes, I am planning to run some latency benchmarks for some of the open llama models once I get access to the gpu. I can also add the measure perplexity metrics when adding the latency results. 
Right now, I don't have all of the numbers but wanted to add the pass to main so we can share the example with collaborators.

File: olive/passes/pytorch/sparsegpt.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        # pytorch_model = copy.deepcopy(model.model)
+        pytorch_model.eval()
+        use_cache = pytorch_model.config.use_cache
+        pytorch_model.config.use_cache = False
+
+        # get module list of layers
+        layers = get_layers(pytorch_model, model_type)
+
+        # get the inputs to the first layer
+        logger.debug("Getting inputs to first layer...")
```
Comment (guotuofeng) [inline olive/passes/pytorch/sparsegpt.py:126]: cleaning these verbose logs?

File: olive/passes/pytorch/sparsegpt.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        logger.debug(f"Running SparseGPT on device: {device}")
+
+        # load_data
+        assert config["data_config"] is not None, "Data config is required for SparseGPT pass."
+        logger.debug("Loading data...")
+        dataloader = self._data_config.to_data_container().create_dataloader(data_root)
+        logger.debug(f"Data loaded. Number of batches: {len(dataloader)}")
+
+        # load model
+        logger.debug("Loading model...")
```
Comment (guotuofeng) [inline olive/passes/pytorch/sparsegpt.py:110]: no need the log since the line 107 already is there

File: olive/passes/pytorch/sparsegpt.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+        # get device to use for computations
+        device = config["compute_device"]
+        if device == "auto":
+            device = "cuda" if torch.cuda.is_available() else "cpu"
+        logger.debug(f"Running SparseGPT on device: {device}")
+
+        # load_data
+        assert config["data_config"] is not None, "Data config is required for SparseGPT pass."
+        logger.debug("Loading data...")
```
Comment (guotuofeng) [inline olive/passes/pytorch/sparsegpt.py:105]: remove line 105 since 101 is already there.

File: docs/source/features/passes/pytorch.md (hunk 2 above)
Comment (jambayk) [reply docs/source/features/passes/pytorch.md:64]: Resolved offline. Will add benchmarks later.

File: olive/passes/pytorch/sparsegpt.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        mode = "unstructured" if isinstance(config["sparsity"], float) else "structured"
+        sparsity = config["sparsity"]
+        n, m = sparsity if mode == "structured" else [0, 0]
+        logger.debug(f"Running SparseGPT with on {model_type} with mode: {mode}, sparsity: {sparsity}")
+
+        # get device to use for computations
+        device = config["compute_device"]
+        if device == "auto":
+            device = "cuda" if torch.cuda.is_available() else "cpu"
+        logger.debug(f"Running SparseGPT on device: {device}")
```
Comment (guotuofeng) [inline olive/passes/pytorch/sparsegpt.py:101]: merge (line 95 and 101) the "Running SparseGPT on" to one line

File: olive/passes/pytorch/sparsegpt_utils.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+# https://github.com/IST-DASLab/sparsegpt
+# https://arxiv.org/abs/2301.00774
+# -------------------------------------------------------------------------
+import logging
+import math
+
+import torch
+import transformers
+
+logger = logging.getLogger(__name__)
```
Comment (guotuofeng) [inline olive/passes/pytorch/sparsegpt_utils.py:15]: not used?

File: olive/passes/pytorch/sparsegpt.py (hunk 3 above)
Comment (jambayk) [reply olive/passes/pytorch/sparsegpt.py]: Done!

File: olive/passes/pytorch/sparsegpt.py (hunk 4 above)
Comment (jambayk) [reply olive/passes/pytorch/sparsegpt.py]: Done!

File: olive/passes/pytorch/sparsegpt.py (hunk 5 above)
Comment (jambayk) [reply olive/passes/pytorch/sparsegpt.py]: Done!

File: olive/passes/pytorch/sparsegpt_utils.py (hunk 7 above)
Comment (jambayk) [reply olive/passes/pytorch/sparsegpt_utils.py]: Done!

**Review by guotuofeng — APPROVED** [review]
